### ISSUE TITLE
addDrugUse includes drug records from before cohort start date.

### ISSUE THREAD

**Issue by ablack3** [darwin-eu-dev/DrugUtilisation#422]
I'm not sure what the intended behaviour of this function is. For the MDD study I'm working on I need to get duration and quantity for a class of drugs (defined by multiple ingredient concepts) that start after the cohort_start date. 

I tried to use addDrugUse but there are two main issues with it.

1. I can only use single ingredient when really what I need is to pass in a concept set. All drug records that match the concept set should be included and treated as "the same" for the purpose of duration and quantity calculations.

2. I want only the first drug that occurs after index. Multiple drug records can be combined if they are within the `eraGap` days of each other. But once someone stops taking a drug from the class the duration calculation should stop.

**Example of a person who has one drug record that begins before index**

The drug record has a missing end date.

``` r
# use latest development version
# remotes::install_github("darwin-eu-dev/DrugUtilisation")

library(CDMConnector)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
con <- DBI::dbConnect(duckdb::duckdb(), eunomia_dir("empty_cdm"))
cdm <- cdm_from_con(con, "main", "main", cdm_name = "test")

person <- tribble(
  ~person_id, ~year_of_birth, ~gender_concept_id,
  1L,         1990,           8507
)

observation_period <- tribble(
  ~observation_period_id, ~person_id, ~observation_period_start_date, ~observation_period_end_date,
  1L, 1L, "2010-01-01", "2022-01-01"
) %>% dplyr::mutate(across(matches("date"), as.Date))

# 3571345 Epilepsy condition
condition_occurrence <- tribble(
  ~condition_occurrence_id, ~person_id, ~condition_concept_id, ~condition_start_date, ~condition_end_date,
  1L, 1L, 3571345L, as.Date("2020-01-01"), as.Date("2021-01-01"),
)

# 1112807 asprin ingredient
drug_exposure <- tribble(
  ~drug_exposure_id, ~person_id, ~drug_concept_id, ~drug_exposure_start_date, ~drug_exposure_end_date, ~quantity,
  1L, 1L, 1112807L, as.Date("2011-01-01"), NA, 100
)

DBI::dbAppendTable(con, "drug_exposure", drug_exposure)
DBI::dbAppendTable(con, "condition_occurrence", condition_occurrence)
DBI::dbAppendTable(con, "person", person)
DBI::dbAppendTable(con, "observation_period", observation_period)

cdm <- generateConceptCohortSet(cdm, conceptSet = list(condition = 3571345), name = "cohort")

library(DrugUtilisation)
cdm$cohort %>%
  addDrugUse(
    ingredientConceptId = 1112807,
    duration = TRUE,
    quantity = TRUE
  ) %>%
  collect() %>%
  mutate_all(as.character) %>%
  tidyr::gather()
#> # A tibble: 13 × 2
#>    key                          value     
#>    <chr>                        <chr>     
#>  1 cohort_definition_id         1         
#>  2 subject_id                   1         
#>  3 cohort_start_date            2020-01-01
#>  4 cohort_end_date              2022-01-01
#>  5 duration                     732       
#>  6 number_exposures             1         
#>  7 cumulative_quantity          100       
#>  8 initial_quantity             100       
#>  9 impute_duration_percentage   100       
#> 10 number_eras                  0         
#> 11 impute_daily_dose_percentage <NA>      
#> 12 initial_daily_dose           <NA>      
#> 13 cumulative_dose              <NA>


cdmDisconnect(cdm)
```

<sup>Created on 2024-03-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

It is not clear to me why duration is 732 (2 years). This is the duration of the condition cohort and not the duration of drug use correct?

If I add an end date to the drug exposure record I still get duration but no longer get any quantity.

``` r
# use latest development version
# remotes::install_github("darwin-eu-dev/DrugUtilisation")

library(CDMConnector)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
con <- DBI::dbConnect(duckdb::duckdb(), eunomia_dir("empty_cdm"))
cdm <- cdm_from_con(con, "main", "main", cdm_name = "test")

person <- tribble(
  ~person_id, ~year_of_birth, ~gender_concept_id,
  1L,         1990,           8507
)

observation_period <- tribble(
  ~observation_period_id, ~person_id, ~observation_period_start_date, ~observation_period_end_date,
  1L, 1L, "2010-01-01", "2022-01-01"
) %>% dplyr::mutate(across(matches("date"), as.Date))

# 3571345 Epilepsy condition
condition_occurrence <- tribble(
  ~condition_occurrence_id, ~person_id, ~condition_concept_id, ~condition_start_date, ~condition_end_date,
  1L, 1L, 3571345L, as.Date("2020-01-01"), as.Date("2021-01-01"),
)

# 1112807 asprin ingredient
drug_exposure <- tribble(
  ~drug_exposure_id, ~person_id, ~drug_concept_id, ~drug_exposure_start_date, ~drug_exposure_end_date, ~quantity,
  1L, 1L, 1112807L, as.Date("2011-01-01"), as.Date("2012-01-01"), 100
)

DBI::dbAppendTable(con, "drug_exposure", drug_exposure)
DBI::dbAppendTable(con, "condition_occurrence", condition_occurrence)
DBI::dbAppendTable(con, "person", person)
DBI::dbAppendTable(con, "observation_period", observation_period)

cdm <- generateConceptCohortSet(cdm, conceptSet = list(condition = 3571345), name = "cohort")

library(DrugUtilisation)
cdm$cohort %>%
  addDrugUse(
    ingredientConceptId = 1112807,
    duration = TRUE,
    quantity = TRUE
  ) %>%
  collect() %>%
  mutate_all(as.character) %>%
  tidyr::gather()
#> # A tibble: 13 × 2
#>    key                          value     
#>    <chr>                        <chr>     
#>  1 cohort_definition_id         1         
#>  2 subject_id                   1         
#>  3 cohort_start_date            2020-01-01
#>  4 cohort_end_date              2022-01-01
#>  5 duration                     732       
#>  6 number_exposures             0         
#>  7 cumulative_quantity          <NA>      
#>  8 initial_quantity             <NA>      
#>  9 impute_duration_percentage   <NA>      
#> 10 number_eras                  0         
#> 11 impute_daily_dose_percentage <NA>      
#> 12 initial_daily_dose           <NA>      
#> 13 cumulative_dose              <NA>


cdmDisconnect(cdm)
```

<sup>Created on 2024-03-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by ablack3** [comment]
All drug records that either start before the index date and have missing drug_exposure_end_date or that overlap with the cohort_start_date are included.
``` r
# use latest development version
# remotes::install_github("darwin-eu-dev/DrugUtilisation")

library(CDMConnector)
library(dplyr, warn.conflicts = F)
con <- DBI::dbConnect(duckdb::duckdb(), eunomia_dir("empty_cdm"))
cdm <- cdm_from_con(con, "main", "main", cdm_name = "test")

person <- tribble(
  ~person_id, ~year_of_birth, ~gender_concept_id,
  1L,         1990,           8507
)

observation_period <- tribble(
  ~observation_period_id, ~person_id, ~observation_period_start_date, ~observation_period_end_date,
  1L, 1L, "2010-01-01", "2022-01-01"
) %>% dplyr::mutate(across(matches("date"), as.Date))

# 3571345 Epilepsy condition
condition_occurrence <- tribble(
  ~condition_occurrence_id, ~person_id, ~condition_concept_id, ~condition_start_date, ~condition_end_date,
  1L, 1L, 3571345L, as.Date("2020-01-01"), as.Date("2021-01-01"),
)

# 1112807 asprin ingredient
drug_exposure <- tribble(
  ~drug_exposure_id, ~person_id, ~drug_concept_id, ~drug_exposure_start_date, ~drug_exposure_end_date, ~quantity,
  1L, 1L, 1112807L, as.Date("2011-01-01"), NA, 100,
  1L, 1L, 1112807L, as.Date("2019-01-01"), as.Date("2021-01-01"), 100
)

DBI::dbAppendTable(con, "drug_exposure", drug_exposure)
DBI::dbAppendTable(con, "condition_occurrence", condition_occurrence)
DBI::dbAppendTable(con, "person", person)
DBI::dbAppendTable(con, "observation_period", observation_period)

cdm <- generateConceptCohortSet(cdm, conceptSet = list(condition = 3571345), name = "cohort")

library(DrugUtilisation)
cdm$cohort %>%
  addDrugUse(
    ingredientConceptId = 1112807,
    duration = TRUE,
    quantity = TRUE
  ) %>%
  collect() %>%
  mutate_all(as.character) %>%
  tidyr::gather()
#> # A tibble: 13 × 2
#>    key                          value     
#>    <chr>                        <chr>     
#>  1 cohort_definition_id         1         
#>  2 subject_id                   1         
#>  3 cohort_start_date            2020-01-01
#>  4 cohort_end_date              2022-01-01
#>  5 duration                     732       
#>  6 number_exposures             2         
#>  7 cumulative_quantity          200       
#>  8 initial_quantity             200       
#>  9 impute_duration_percentage   50        
#> 10 number_eras                  1         
#> 11 impute_daily_dose_percentage 100       
#> 12 initial_daily_dose           <NA>      
#> 13 cumulative_dose              <NA>


cdmDisconnect(cdm)
```

<sup>Created on 2024-03-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

why is number_eras 1 in this example above but 0 in the first example where the drug exposure starts before the cohort_start_date?

**Comment by ablack3** [comment]
more tests...

``` r
# use latest development version
# remotes::install_github("darwin-eu-dev/DrugUtilisation")

run_test <- function(drug_exposures) {
  library(CDMConnector)
  library(dplyr, warn.conflicts = F)
  con <- DBI::dbConnect(duckdb::duckdb(), eunomia_dir("empty_cdm"))
  cdm <- cdm_from_con(con, "main", "main", cdm_name = "test")

  person <- tribble(
    ~person_id, ~year_of_birth, ~gender_concept_id,
    1L,         1990,           8507
  )

  observation_period <- tribble(
    ~observation_period_id, ~person_id, ~observation_period_start_date, ~observation_period_end_date,
    1L, 1L, "2010-01-01", "2022-01-01"
  ) %>% dplyr::mutate(across(matches("date"), as.Date))

  # 3571345 Epilepsy condition
  condition_occurrence <- tribble(
    ~condition_occurrence_id, ~person_id, ~condition_concept_id, ~condition_start_date, ~condition_end_date,
    1L, 1L, 3571345L, as.Date("2020-01-01"), as.Date("2021-01-01"),
  )

  # # 1112807 asprin ingredient
  # drug_exposure <- tribble(
  #   ~drug_exposure_id, ~person_id, ~drug_concept_id, ~drug_exposure_start_date, ~drug_exposure_end_date, ~quantity,
  #   # 1L, 1L, 1112807L, as.Date("2020-04-01"), NA, 100,
  #   1L, 1L, 1112807L, as.Date("2020-05-01"), as.Date("2020-06-01"), 100,
  #   1L, 1L, 1112807L, as.Date("2020-07-01"), as.Date("202-08-01"), 100
  # )

  DBI::dbAppendTable(con, "drug_exposure", drug_exposure)
  DBI::dbAppendTable(con, "condition_occurrence", condition_occurrence)
  DBI::dbAppendTable(con, "person", person)
  DBI::dbAppendTable(con, "observation_period", observation_period)

  cdm <- generateConceptCohortSet(cdm, conceptSet = list(condition = 3571345), name = "cohort")

  library(DrugUtilisation)
  cdm$cohort %>%
    addDrugUse(
      ingredientConceptId = 1112807,
      duration = TRUE,
      quantity = TRUE,
      gapEra = 0
    ) %>%
    collect() %>%
    mutate_all(as.character) %>%
    tidyr::gather() %>%
    print()


  cdmDisconnect(cdm)

}

# in this case only the first record is used
drug_exposure <- tribble(
  ~drug_exposure_id, ~person_id, ~drug_concept_id, ~drug_exposure_start_date, ~drug_exposure_end_date, ~quantity,
  1L, 1L, 1112807L, as.Date("2020-05-01"), as.Date("2020-06-01"), 100,
  1L, 1L, 1112807L, as.Date("2020-07-01"), as.Date("202-08-01"), 100
)
#> Error in tribble(~drug_exposure_id, ~person_id, ~drug_concept_id, ~drug_exposure_start_date, : could not find function "tribble"

run_test(drug_exposure)
#> Error in run_test(drug_exposure): object 'drug_exposure' not found

# adding a record without an end date first will capture the first two records
drug_exposure <- tribble(
  ~drug_exposure_id, ~person_id, ~drug_concept_id, ~drug_exposure_start_date, ~drug_exposure_end_date, ~quantity,
  1L, 1L, 1112807L, as.Date("2020-04-01"), NA, 100,
  1L, 1L, 1112807L, as.Date("2020-05-01"), as.Date("2020-06-01"), 100,
  1L, 1L, 1112807L, as.Date("2020-07-01"), as.Date("202-08-01"), 100
)

run_test(drug_exposure)
#> # A tibble: 13 × 2
#>    key                          value     
#>    <chr>                        <chr>     
#>  1 cohort_definition_id         1         
#>  2 subject_id                   1         
#>  3 cohort_start_date            2020-01-01
#>  4 cohort_end_date              2022-01-01
#>  5 duration                     732       
#>  6 number_exposures             2         
#>  7 cumulative_quantity          200       
#>  8 initial_quantity             <NA>      
#>  9 impute_duration_percentage   50        
#> 10 number_eras                  1         
#> 11 impute_daily_dose_percentage 100       
#> 12 initial_daily_dose           <NA>      
#> 13 cumulative_dose              <NA>


# in this case we also capture both records even though the era gap is 0
drug_exposure <- tribble(
  ~drug_exposure_id, ~person_id, ~drug_concept_id, ~drug_exposure_start_date, ~drug_exposure_end_date, ~quantity,
  1L, 1L, 1112807L, as.Date("2020-04-01"), as.Date("2020-05-01"), 100,
  1L, 1L, 1112807L, as.Date("2020-07-01"), NA, 100
)

run_test(drug_exposure)
#> # A tibble: 13 × 2
#>    key                          value     
#>    <chr>                        <chr>     
#>  1 cohort_definition_id         1         
#>  2 subject_id                   1         
#>  3 cohort_start_date            2020-01-01
#>  4 cohort_end_date              2022-01-01
#>  5 duration                     732       
#>  6 number_exposures             2         
#>  7 cumulative_quantity          200       
#>  8 initial_quantity             <NA>      
#>  9 impute_duration_percentage   50        
#> 10 number_eras                  1         
#> 11 impute_daily_dose_percentage 100       
#> 12 initial_daily_dose           <NA>      
#> 13 cumulative_dose              <NA>



# in this case we also capture both records even though the era gap is 0
drug_exposure <- tribble(
  ~drug_exposure_id, ~person_id, ~drug_concept_id, ~drug_exposure_start_date, ~drug_exposure_end_date, ~quantity,
  1L, 1L, 1112807L, as.Date("2020-04-01"), as.Date("2020-05-01"), 100,
  1L, 1L, 1112807L, as.Date("2020-07-01"), NA, 100
)

run_test(drug_exposure)
#> # A tibble: 13 × 2
#>    key                          value     
#>    <chr>                        <chr>     
#>  1 cohort_definition_id         1         
#>  2 subject_id                   1         
#>  3 cohort_start_date            2020-01-01
#>  4 cohort_end_date              2022-01-01
#>  5 duration                     732       
#>  6 number_exposures             2         
#>  7 cumulative_quantity          200       
#>  8 initial_quantity             <NA>      
#>  9 impute_duration_percentage   50        
#> 10 number_eras                  1         
#> 11 impute_daily_dose_percentage 100       
#> 12 initial_daily_dose           <NA>      
#> 13 cumulative_dose              <NA>
```

<sup>Created on 2024-03-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by ablack3** [comment]
So it seems some of this unexpected behavior is coming from missing end dates. But also I would expect that `duration` refers to time on treatment. We definitely need drug exposure duration for studies I think.

**Comment by catalamarti** [comment]
Hi @ablack3 so the function has this behavior on purpose.
addDrugUse is mean to be used to add the dosage and drugs between cohort_start_date and cohort_end_date, if prescriptions started before start date but continue during the cohort start date the contributed dosage inside during the exposed period is considered.
This was discussed at the initial stages of DrugUtilisation and the users seems that they wanted this behavior. I am happy to discuss and potentially add a new argument if there are use cases where this is needed.
The study period of the person is between cohort_start_date and cohort_end_date, so thats why we consider other drug eras/exposures even if the treatment stops

**Comment by vperaEMC** [comment]
I can relate to Adam's first point.

>     1. I can only use single ingredient when really what I need is to pass in a concept set. All drug records that match the concept set should be included and treated as "the same" for the purpose of duration and quantity calculations.
> 

For now, I am specifically interested in calculating the drug duration of every single row in my drugutilizers cohort table and create a summary statistic on this. A simple workaround was calculating the date difference between cohort start and end date, and adding these values as a new column in the cohort table. Using the summeriseResult function furthered helped me get the appropriate summary statistics. (Thanks for the help, @ablack3 !)

**Comment by ablack3** [comment]
> This was discussed at the initial stages of DrugUtilisation and the users seems that they wanted this behavior. I am happy to discuss and potentially add a new argument if there are use cases where this is needed.

To me it was quite confusing that duration comes from the input cohort and quantity comes from the ingredient input. Yea I think this function needs some discussion.



> The study period of the person is between cohort_start_date and cohort_end_date, so thats why we consider other drug eras/exposures even if the treatment stops

Not necessarily for all studies. In my case I had a cohort that had nothing to do with drugs and wanted to get drug use during specific time windows relative to index, not necessarily only between the cohort start and end dates. 

What I needed was very similar to addConceptIntersect in patient profiles but with the additional attributes of quantity and duration of drug use. so it almost seems like the general approach is addConceptIntersect, but in the case of drugs there are additional attributes that we may want to add. But addDrugUse should work very similarly to addConceptIntersect. That's my two cents after using it anyway.

**Comment by catalamarti** [comment]
Mmmmmm so I see both things being interesting to add!
For the moment this function (and basically the whole package) it is narrowed to fulfill what it is needed for the DUS as defined in DARWIN-EU Catalog of Standard Analytics. But I am happy to wide a bit the scope and include all these functionalities that seem useful!
We are planning a sprint end April / beginning of May for DrugUtilisation so would be good to discuss this before that

**Comment by catalamarti** [comment]
this had been implemented in addDrugUtilisation function #468
